Update `validate` and `versions`, refactor `metadata` and `urls`

johnmhoran opened the pull request:
Reference: https://github.com/nexB/purldb/issues/247

johnmhoran commented:
@JonoYang Please review when you have the chance.  Looking forward to your suggestions and questions.  This comprises my work (subject to input from you and colleagues) on the first four commands: `metadata`, `urls`, `validate` and `versions`.  

I've substantially updated, refactored and supplemented `purlcli.py`, `test_purlcli.py` and the related test files.  All `make test` tests passed, all GH checks passed, no apparent merge conflict with `main` per GH.

JonoYang commented on purldb-toolkit/src/purldb_toolkit/purlcli.py:
```diff
@@ -123,21 +112,65 @@ def get_metadata_details(purls, output, file, unique, command_name):
+        return "not_in_upstream_repo"
+
+
+def normalize_purls(purls, unique):
```
Add a docstring explaining this function

johnmhoran replied:
Docstring added for `normalize_purls()`.

JonoYang commented on purldb-toolkit/src/purldb_toolkit/purlcli.py:
```diff
@@ -185,7 +219,7 @@ def construct_headers(
 
-    if command_name == "metadata" and unique:
+    if (command_name in ["metadata", "urls", "validate", "versions"]) and unique:
         for purl in normalized_purls:
```
```suggestion
        for input_purl, normalized_purl in normalized_purls:
            if input_purl != normalized_purl:
                warnings.append(f"input PURL: '{input_purl}' normalized to '{normalized_purl}'")
```

johnmhoran replied:
Done.  Thanks @JonoYang -- naming the pair of PURLs is a nice clarification and I'll remember this going forward.

JonoYang commented on purldb-toolkit/src/purldb_toolkit/purlcli.py:
```diff
@@ -194,38 +228,27 @@ def construct_headers(
+        if command_name in ["metadata", "urls", "validate", "versions"]:
+            purl_warning = purl_warnings.get(purl, None)
+            if purl_warning:
+                warnings.append(warning_text[purl_warning])
```
```suggestion
                warning = warning_text[purl_warning]
                warnings.append(warning)
                print(warning)
```

johnmhoran replied:
Another good clarification -- done.

JonoYang commented on purldb-toolkit/src/purldb_toolkit/purlcli.py:
```diff
@@ -346,21 +333,17 @@ def get_urls_details(purls, output, file, head, command_name):
 
         urls_purl = check_urls_purl(purl)
```
I think `urls_purl` should be renamed `purl_status` or something that would indicate what the return value is

johnmhoran replied:
Another good clarification -- done.

JonoYang commented on purldb-toolkit/src/purldb_toolkit/purlcli.py:
```diff
@@ -623,64 +703,86 @@ def get_versions(purls, output, file):
             purl_version_data = {}
-            purl_version = package_version_object.to_dict()["value"]
-            nested_purl = purl + "@" + f"{purl_version}"
+            purl_version = package_version.to_dict()["value"]
```
I think you can do this:

```suggestion
            purl_version = package_version.value
```

JonoYang commented on purldb-toolkit/src/purldb_toolkit/purlcli.py:
```diff
@@ -623,64 +703,86 @@ def get_versions(purls, output, file):
             purl_version_data["version"] = f"{purl_version}"
             purl_version_data["release_date"] = (
-                f'{package_version_object.to_dict()["release_date"]}'
+                f'{package_version.to_dict()["release_date"]}'
```
```suggestion
                f'{package_version.release_date}'
```

johnmhoran replied:
@JonoYang Very interesting.  It took me a while to figure out why we could replace `package_version.to_dict()["value"]` with a dot approach -- `package_version.value`.  I saw nothing in `purlcli.py` itself that gave me a hint (`from fetchcode.package_versions import SUPPORTED_ECOSYSTEMS, versions` was eventually an indirect hint).  

I finally found that in fetchcode's `package_versions.py` we have a `@dataclasses.dataclass` decorator that defines the `PackageVersion` class -- a class whose name I'd previously noticed in a console error message when I tried modifying the name after the dot, wondering how this worked.  

I see the `value` and `release_date` attributes defined in that class, but how is it that I can use that class in `purlcli.py` without importing the class (and we don't, at least not by name)?

johnmhoran replied:
With the change, 4 tests now fail ....

JonoYang replied:
@johnmhoran 

> I see the value and release_date attributes defined in that class, but how is it that I can use that class in purlcli.py without importing the class (and we don't, at least not by name)?

At the top of `purlcli.py`, you have the import statement `from fetchcode.package_versions import SUPPORTED_ECOSYSTEMS, versions` . Since `PackageVersion` is defined in `fetchcode.package_versions`, that is also imported

johnmhoran replied:
Done.  The test failures were caused when some, but not all, of the generated `release_date` values used a slightly different format, e.g.,

```
'2013-12-11T00:27:10.097000+00:00'
vs.
'2013-12-11 00:27:10.097000+00:00'

```

JonoYang commented on purldb-toolkit/src/purldb_toolkit/purlcli.py:
```diff
@@ -194,38 +228,27 @@ def construct_headers(
-        if command_name == "urls" and urls_purl == "not_in_upstream_repo":
-            warnings.append(f"'{purl}' does not exist in the upstream repo")
-            continue
+    log_file = Path("purldb-toolkit/src/purldb_toolkit/app.log")
```
I think the log file should be saved in the users home directory, or some other canonical place for log files.

johnmhoran replied:
@JonoYang I'm encountering syntax errors trying to save the `app.log` in the user's home directory.  Variations on 

```
filename = os.path.join(os.path.expanduser("~"), 'app.log')
```

keep returning 

```
    filename = os.path.join(os.path.expanduser("~"), 'app.log')
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
SyntaxError: invalid syntax. Perhaps you forgot a comma?
```

I've explored a number of stackoverflow and similar hints (nothing re saving but examples of finding the user's home and expanding, which should be applicable) but no luck yet.  Do you know what the proper syntax is?

JonoYang replied:
@johnmhoran I'm not sure what the issue is. I'm able to run `filename = os.path.join(os.path.expanduser("~"), 'app.log')` in the Python shell. Maybe you can try `filename = Path.home() / 'app.log'`

johnmhoran replied:
Thanks @JonoYang -- I got it working in the users home directory a while ago -- user error.

```
        logging.basicConfig(
            filename=os.path.join(os.path.expanduser("~"), "app.log"),
            level=logging.ERROR,
            format="%(levelname)s - %(message)s",
            filemode="w",
        )

        logger.error(f"validate_purl(): json.decoder.JSONDecodeError for '{purl}': {e}")
```

JonoYang reviewed (changes requested):
@johnmhoran I have some suggestions for you.

johnmhoran commented:
@JonoYang I've addressed all of your comments and committed and pushed; all GH checks passed (I had to rerun a 3.10 push check that failed initially -- transitory validation warning for one of the PURLs in one of the tests).

johnmhoran commented:
@JonoYang I'm taking a closer look at this, which doesn't look quite right.  Maybe a copy-paste error on my part?  Unable to find the possible source for that approach.  This appears twice in `purlcli.py`.

```
log_file = Path(log_file=os.path.join(os.path.expanduser("~"), "app.log"))
```

Maybe it should be

```
log_file = Path(os.path.join(os.path.expanduser("~"), "app.log"))
```

Testing manually....

johnmhoran commented:
@JonoYang Unable so far to manually trigger a `pkg:deb/debian/*` error, which is the only way I know to test the try/except.  ;-)  Will keep trying because I'm sure that example had excess text (which I've removed and am testing).  The error will surely occur again ....

johnmhoran commented:
@JonoYang I've replicated two rounds of the error followed by no error and see the `app.log` file gets created (in my Ubuntu-20.04 `home/jmh/` directory) and populated properly and deleted in the next round as intended.  I'll take care of running the main tests, commit and push this correction, and watch the GH checks run.

johnmhoran commented:
@JonoYang All GH checks passed.

johnmhoran commented on purldb-toolkit/src/purldb_toolkit/purlcli.py:
```diff
@@ -185,47 +226,39 @@ def construct_headers(
-        if command_name == "urls" and urls_purl == "not_in_upstream_repo":
-            warnings.append(f"'{purl}' does not exist in the upstream repo")
-            continue
+    log_file = Path(os.path.join(os.path.expanduser("~"), "app.log"))
```
Thank you @JonoYang -- I'll implement these improvements shortly.  👍

johnmhoran commented:
@JonoYang Changes implemented, just one local failure with `make test` -- the usual `minecode/tests/test_maven.py::MavenEnd2EndTest::test_visit_and_map_with_index` -- and all GH checks passed.

JonoYang commented:
Thanks! Merging this now
